### PR TITLE
Add Emformer RNN-T model

### PR DESCRIPTION
**Pull request by hwangjeff**
Adds streaming-capable recurrent neural network transducer (RNN-T) model that uses Emformer for its transcription network. Includes two factory functions — one that allows for building a custom model, and one that builds a preconfigured base model.

### CODE REVIEW

File: test/torchaudio_unittest/prototype/rnnt_test_impl.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+            "max_input_length": 61,
+            "num_symbols": self._get_model_config()["num_symbols"],
+            "max_target_length": 23,
+            "input_dim": self._get_model_config()["input_dim"],
+            "right_context_length": self._get_model_config()["right_context_length"],
+            "encoding_dim": self._get_model_config()["encoding_dim"],
+            "joiner_max_input_length": 61
+            // self._get_model_config()["time_reduction_stride"],
+            "segment_length": self._get_model_config()["segment_length"],
+            "time_reduction_stride": self._get_model_config()["time_reduction_stride"],
```
Comment (mthrok) [inline test/torchaudio_unittest/prototype/rnnt_test_impl.py:19]: Looks like the result of `self._get_model_config()` can be cached and re-used.

File: torchaudio/prototype/rnnt.py [hunk 2]
```diff
@@ -0,0 +1,786 @@
...
+        lstm_layer_norm=lstm_layer_norm,
+        lstm_layer_norm_epsilon=lstm_layer_norm_epsilon,
+        lstm_dropout=lstm_dropout,
+    )
+    joiner = _Joiner(encoding_dim, num_symbols)
+    return _RNNT(transcriber, predictor, joiner)
+
+
+def emformer_rnnt_base() -> _RNNT:
+    r"""Builds basic version of Emformer RNN-T model.
```
Comment (mthrok) [inline torchaudio/prototype/rnnt.py:758]: Is there a corresponding name referred in the original paper?
In https://arxiv.org/abs/2010.10759, I see architectures referred as `EM-24L` and `EM-36L`.

I assume that these factory functions follow the pattern of Wav2Vec2 module. In the case of wav2vec2, the model architectures are referred to as base/large etc in the paper.

File: torchaudio/prototype/rnnt.py [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+
+    Returns:
+        _RNNT:
```
Comment (mthrok) [inline torchaudio/prototype/rnnt.py:761]: Since the returned model is part of public API, `_RNNT` needs to be public as well.

**Review by mthrok — COMMENTED** [review]
Could you add the new models to the documentation?

https://github.com/pytorch/audio/blob/main/docs/source/prototype.rst

File: torchaudio/prototype/rnnt.py (hunk 2 above)
Comment (hwangjeff) [reply torchaudio/prototype/rnnt.py:758]: no, there isn't

File: torchaudio/prototype/rnnt.py [hunk 4]
```diff
@@ -0,0 +1,786 @@
...
+        T_max = num_frames // self.stride
+
+        output = input.reshape(B, T_max, D * self.stride)
+        output = output.contiguous()
+        return output, lengths
+
+
+class _CustomLSTM(torch.nn.Module):
+    r"""Custom long-short-term memory (LSTM) block that additionally applies
+    linear projection and layer normalization to internal nodes.
```
Comment (nateanl) [inline torchaudio/prototype/rnnt.py:58]: Would be great to describe the details in the `_CustomLSTM`. If I understand correctly, the difference than conventional LSTM is that the bias of layer `x2g` (if layer norm is applied) and `p2g` are removed, The input gate `i_t` and cell gate  `c_t` are layer-normed?

Technically it's not `additionally linear projection` since LSTM also uses linear weights (but with bias) for input gates and hidden states, but an equation or a description will fully explain it.

File: torchaudio/prototype/rnnt.py [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+            outputs.append(h)
+
+        output = torch.stack(outputs, dim=0)
+        state = [h, c]
+
+        return output, state
+
+
+class _Transcriber(torch.nn.Module):
+    r"""Recurrent neural network transducer (RNN-T) transcription network.
```
Comment (nateanl) [inline torchaudio/prototype/rnnt.py:138]: This is more like an Emformer component of RNN-T transcription network :)?

File: torchaudio/prototype/rnnt.py [hunk 6]
```diff
@@ -0,0 +1,786 @@
...
+                    layer_norm=lstm_layer_norm,
+                    layer_norm_epsilon=lstm_layer_norm_epsilon,
+                )
+                for idx in range(num_lstm_layers)
+            ]
+        )
+        self.linear = torch.nn.Linear(symbol_embedding_dim, output_dim)
+        self.output_layer_norm = torch.nn.LayerNorm(output_dim)
+
+        self.lstm_dropout = lstm_dropout
```
Comment (nateanl) [inline torchaudio/prototype/rnnt.py:329]: A module-level should be more standard? 
`self.dropout = torch.nn.Dropout (p=lstm_dropout)`

File: torchaudio/prototype/rnnt.py [hunk 7]
```diff
@@ -0,0 +1,786 @@
...
+        input_tb = input.permute(1, 0)
+        embedding_out = self.embedding(input_tb)
+        input_layer_norm_out = self.input_layer_norm(embedding_out)
+
+        lstm_out = input_layer_norm_out
+        state_out: List[List[torch.Tensor]] = []
+        for layer_idx, lstm in enumerate(self.lstm_layers):
+            lstm_out, lstm_state_out = lstm(
+                lstm_out, None if state is None else state[layer_idx]
+            )
```
Comment (nateanl) [inline torchaudio/prototype/rnnt.py:372]: The logic here doesn't look correct. `lstm_state_out` is not updated as `lstm_out`?

File: torchaudio/prototype/rnnt.py [hunk 8]
```diff
@@ -0,0 +1,786 @@
...
+            target_encodings (torch.Tensor): target encoding sequences, with shape `(B, U, D)`.
+            target_lengths (torch.Tensor): with shape `(B,)` and i-th element representing
+                valid sequence length of i-th batch element in ``target_encodings``.
+
+        Returns:
+            (torch.Tensor, torch.Tensor, torch.Tensor):
+                torch.Tensor
+                    joint network output, with shape `(B, T, U, D)`.
+                torch.Tensor
+                    output input lengths, with shape `(B,)` and i-th element representing
```
Comment (nateanl) [inline torchaudio/prototype/rnnt.py:423]:
```suggestion
                    output lengths, with shape `(B,)` and i-th element representing
```

File: torchaudio/prototype/rnnt.py [hunk 9]
```diff
@@ -0,0 +1,786 @@
...
+                    number of valid elements along dim 1 for i-th batch element in joint network output.
+                torch.Tensor
+                    output target lengths, with shape `(B,)` and i-th element representing
+                    number of valid elements along dim 2 for i-th batch element in joint network output.
+        """
+        joint_encodings = (
+            input_encodings.unsqueeze(2).contiguous()
+            + target_encodings.unsqueeze(1).contiguous()
+        )
+        relu_out = torch.nn.functional.relu(joint_encodings)
```
Comment (nateanl) [inline torchaudio/prototype/rnnt.py:433]: Same here, we can use Module for ReLU :)

File: torchaudio/prototype/rnnt.py (hunk 4 above)
Comment (hwangjeff) [reply torchaudio/prototype/rnnt.py:58]: yeah the language could be more precise — will revise

File: torchaudio/prototype/rnnt.py (hunk 5 above)
Comment (hwangjeff) [reply torchaudio/prototype/rnnt.py:138]: for now, yeah. i anticipate once we add in other transformer flavors that we can generalize this beyond emformer

File: torchaudio/prototype/rnnt.py (hunk 7 above)
Comment (hwangjeff) [reply torchaudio/prototype/rnnt.py:372]: can you elaborate on this?

File: torchaudio/prototype/rnnt.py (hunk 8 above)
Comment (hwangjeff) [reply torchaudio/prototype/rnnt.py:423]: i added the "input" bit with the intent to clarify that these output lengths correspond to `input`. perhaps i can change `input` to something like `utterance` or `source` to reduce the ambiguity? what do you think?

File: torchaudio/prototype/rnnt.py (hunk 7 above)
Comment (nateanl) [reply torchaudio/prototype/rnnt.py:372]: The `lstm_out` of the last layer is used as the input for the next layer, while ``lstm_state_out`` is not used in this way. I'm wondering if it's expected.

File: torchaudio/prototype/rnnt.py (hunk 8 above)
Comment (nateanl) [reply torchaudio/prototype/rnnt.py:423]: Yeah that's a good idea.

File: torchaudio/prototype/rnnt.py (hunk 7 above)
Comment (hwangjeff) [reply torchaudio/prototype/rnnt.py:372]: got it. yeah this is expected — we maintain state independently on a per-layer basis and provide each layer with the state it generated in the previous timestep

File: torchaudio/prototype/rnnt.py [hunk 10]
```diff
@@ -0,0 +1,786 @@
...
+            source_encodings.unsqueeze(2).contiguous()
+            + target_encodings.unsqueeze(1).contiguous()
+        )
+        relu_out = self.relu(joint_encodings)
+        output = self.linear(relu_out)
+        return output, source_lengths, target_lengths
+
+
+class RNNT(torch.nn.Module):
+    r"""Recurrent neural network transducer (RNN-T) model.
```
Comment (mthrok) [inline torchaudio/prototype/rnnt.py:439]: https://435773-90321822-gh.circle-artifacts.com/0/docs/prototype.html#torchaudio.prototype.RNNT

Can you add a note to redirect users to the factory functions? Also the doctoring referring to private classes is strange, can you replace them with `torch.nn.Module`? (I would change type annotation as well but that might break TorchScript, so that's fine.)

ref: https://pytorch.org/audio/stable/models.html#wav2vec2model

**Review by mthrok — APPROVED** [review]
